warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/facet/FacetHandler.java

```diff
@@ -0,0 +1,366 @@
+                        Key result = new Key(row, myCf, dateColumnQualifier, cv, event.getDate());
+                        results.put(new BulkIngestKey(facetTableName, result), sharedValue);
+                        
+                        countWritten++;
+                    }
+                    
+                    myCf = generatedCf;
+                }
+                
+                Text row = createFieldValuePair(pivotTypes.getIndexedFieldValue(), pivotTypes.getIndexedFieldValue(), event.getDataType());
+                Key result = new Key(row, reflexiveCf, dateColumnQualifier, cv, event.getDate());
+                results.put(new BulkIngestKey(facetTableName, result), sharedValue);
+            }
+        }
+        
+        for (Map.Entry<String,String> pivot : pivotMap.entries()) {
+            Key result = new Key(new Text(pivot.getKey() + NULL + pivot.getValue()), PV);
+            results.put(new BulkIngestKey(facetMetadataTableName, result), EMPTY_VALUE);
+            countWritten++;
+        }
+        contextWriter.write(results, context);
+        return countWritten;
+    }
+    
+    private Multimap<String,NormalizedContentInterface> hashEventFields(Multimap<String,NormalizedContentInterface> fields,
+                    HashTableFunction<KEYIN,KEYOUT,VALUEOUT> func) {
+        Multimap<String,NormalizedContentInterface> eventFields = HashMultimap.create();
+        for (Map.Entry<String,Collection<NormalizedContentInterface>> entry : fields.asMap().entrySet()) {
+            Collection<NormalizedContentInterface> coll = func.apply(entry.getValue());
+            if (coll != null && !coll.isEmpty()) {
+                eventFields.putAll(entry.getKey(), coll);
+            }
+        }
+        return eventFields;
+    }
+    
+    /**
+     * Create the column qualifier that includes pivotFieldValue, facetFieldValue and datatype.
+     * 
+     * @param pivotFieldValue
+     * @param facetFieldValue
+     * @param dataType
+     * @return
+     */
+    protected Text createFieldValuePair(String pivotFieldValue, String facetFieldValue, Type dataType) {
+        return new Text(pivotFieldValue + NULL + facetFieldValue + NULL + dataType.typeName());
+    }
+    
+    /**
+     * Create the column family consisting of pivotFieldName and facetFieldName
+     * 
+     * @param pivotFieldName
+     * @param facetFieldName
+     * @return
+     */
+    protected Text createColumnFamily(String pivotFieldName, String facetFieldName) {
+        return new Text(pivotFieldName + NULL + facetFieldName);
+    }
+    
+    /** TODO */
+    @Override
+    public FacetValue estimate(RawRecordContainer input) {
+        // precision value: 10, sparse set disabled.
+        HyperLogLogPlus card = new HyperLogLogPlus(10);
+        card.offer(input.getId().toString());
+        
+        return new FacetValue(card, new CountMinSketch(10, 1, 1));
```

Comment:
I might not be reading this correctly, but using this constructor we're creating a CMS with depth of 10 and width of 1.  We should probable make these parameters configurable, likewise for hLL++.  That way one could set the parameters based on the error and probability they want to support.  Could add some bounds checking to make it reasonable.

Reply:
Good point. I'll work on the plumbing for this. I'm wondering if we might want to do this on a per-field basis or along some other dimension than "It works this way for everything". I'm also wondering if these parameters get encoded in the serialized object so that we can track this for individual estimates.